packages/astro/components/Debug.astro

```diff
@@ -1,6 +1,308 @@
+  );
+}
+
+const Empty = Symbol('Empty');
+
+const KeyValue = ({ key, value, dim }) => {
+  let type = key === '__proto__' ? 'prototype' : getType(value);
+  if (type === 'null') {
+    value = 'null';
+  } else if (type === 'undefined') {
+    value = 'undefined';
+  } else if (value === Empty) {
+    type = 'empty';
+    value = 'Empty';
+  } else {
+    value = JSON.stringify(value);
+  }
+
+  return (
+    <div class="line">
+      {(key || !key && key === 0) && <><span class={`key ${dim ? 'key-dim' : ''}`.trim()}>{key}</span><span class="sep">:</span></>}
+      <span class={`value value-${type}`}>
+        {value}
+      </span>
+    </div>
+  )
+}
+
+const Node = ({ node, key, ...props }) => {
+  const type = getType(node);
+
+  if (type === 'array' || type === 'object') {
+    let children = [];
+    if (type === 'array' && node.length > 0 && Object.entries(node).length === 0) {
+      children = Array.from({ length: node.length }, (_, key) => <Node node={Empty} key={key} />);
+    } else {
+      children = Object.entries(node).map(([key, value]) => <Node node={value} key={key} />);
+    }
+    return (
+      <Details node={node} key={key} children={children} />
+    );
+  } else if (type === 'function') {
+    return (
+      <Details node={node} key={key} children={
+      <>
+        <KeyValue key="name" value={node.name} dim={true} />
+        <KeyValue key="__proto__" value="Function" dim={true} />
+      </>
+      }/>
+    );
+  } else if (type === 'promise') {
+    return (
+      <Details node={node} key={key} children={
+        <KeyValue key="__proto__" value="Promise" dim={true} />
+      } />
+    );
+  }
+
+  return <KeyValue key={key} value={node} />;
+}
 ---
 
-<!-- JSON.stringify can pretty print our input: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_space_argument -->
-<Prism lang="json" code={JSON.stringify(Astro.props, null, 2)}/>
+<div class="debug">
+  <header>
+    <h2><span class="debug-title">Debug</span> <span class="debug-name">"{key}"</span></h2>
+  </header>
+
+  <main>
+    <Node node={value} />
+  </main>
+</div>
+
+<style lang="scss">
+header {
+  background: #FF1639;
+  margin: -1rem -1.5rem 1rem;
+  padding: 0.25rem 0.75rem;
+}
+
+h2 {
+  font-size: 1em;
+  color: #fff;
+}
+
+.debug-title {
+  font-weight: bold;
+  text-transform: uppercase;
+  margin-right: 0.75em;
+}
+
+.debug {
+  all: initial;
+  display: flex;
+  flex-direction: column;
+  padding: 1rem 1.5rem;
+  overflow-y: hidden;
+  overflow-x: auto;
+  border: 1px solid #FFCFD6;
+
+  background: #FFF;
+  font-family: Consolas, "Andale Mono WT", "Andale Mono", "Lucida Console", "Lucida Sans Typewriter", "DejaVu Sans Mono", "Bitstream Vera Sans Mono", "Liberation Mono", "Nimbus Mono L", Monaco, "Courier New", Courier, monospace;
+  font-size: 0.8rem;
+  line-height: 1.44;
+  color: #6b7280;
+  white-space: pre;
+
+  :global(details[open] > summary span.hide) {
```

Comment:
do all of these need to be global? Would be great to scope, if possible

Reply:
Fixed! Had to prop-drill a bit, but the styles are much cleaner.